File: dart/lib/src/packets/src/trips/ts.dart
part of '../../packets.dart';

class TsPacket extends ClientPacket {
  /// [timestamp] is the time of the packet.
  /// This is identified in the packet as `UNIX`
  final DateTime timestamp;

  /// [tripId] is the trip identifier of the packet.
  /// This is identified in the packet as `TRIP_ID`
  final String tripId;

  /// [TsPacket] is the Trip start packet.
  ///
  /// This packet is part of the packet sent between Layrz services to identify trips.
  TsPacket({
    required this.timestamp,
    required this.tripId,
  });

  /// [fromPacket] creates a [TsPacket] from a string packet in the format of `Layrz Protocol v3`.
  static TsPacket fromPacket(String raw) {
    if (!raw.startsWith('<Ts>') || !raw.endsWith('</Ts>')) {
      throw ParseException('Invalid identification packet, should be <Ts>...</Ts>');
    }

    final parts = raw.substring(4, raw.length - 5).split(';');
    if (parts.length != 3) {
      throw MalformedException('Invalid packet parts, should have 3 parts');
    }

    int? receivedCrc = int.tryParse(parts[2], radix: 16);
    int? calculatedCrc = calculateCrc("${parts.sublist(0, 2).join(';')};".codeUnits);

    if (receivedCrc != calculatedCrc) {
      throw CrcException('Invalid CRC, received: $receivedCrc, calculated: $calculatedCrc');
    }

    DateTime timestamp;
    try {
      timestamp = DateTime.fromMillisecondsSinceEpoch(int.parse(parts[0]) * 1000);
    } catch (e) {
      throw MalformedException('Invalid timestamp');
    }

    return TsPacket(
      timestamp: timestamp,
      tripId: parts[1],
    );
  }

  /// [toPacket] returns the packet in the format of `Layrz Protocol v3`.
  @override
  String toPacket() {
    String payload = '${(timestamp.millisecondsSinceEpoch / 1000).round()};';
    payload += '$tripId;';
    String crc = calculateCrc(payload.codeUnits).toRadixString(16).padLeft(4, '0').toUpperCase();

    return '<Ts>$payload$crc</Ts>';
  }

  @override
  String toString() => toPacket();

  @override
  TsPacket copyWith({
    DateTime? timestamp,
    String? tripId,
  }) {
    return TsPacket(
      timestamp: timestamp ?? this.timestamp,
      tripId: tripId ?? this.tripId,
    );
  }
}

File: go/packets/trips/ts.go
package trips

import (
	"errors"
	"fmt"
	"strconv"
	"strings"
	"time"

	"github.com/goldenm-software/layrz-protocol/go/v3/internal/wire"
)

// TsPacket is the Trip Start packet sent between Layrz services to identify trips.
type TsPacket struct {
	// Timestamp of the packet
	Timestamp time.Time

	// TripId is the unique trip identifier (UUID string)
	TripId string
}

// FromPacket converts a raw <Ts>...</Ts> string to a TsPacket.
func (p *TsPacket) FromPacket(raw *string) error {
	if !strings.HasPrefix(*raw, "<Ts>") || !strings.HasSuffix(*raw, "</Ts>") {
		return errors.New("invalid packet, should be <Ts>...</Ts>")
	}

	body := (*raw)[4 : len(*raw)-5]

	rawCrc := body[len(body)-4:]
	body = body[:len(body)-4]

	receivedCrc, err := strconv.ParseUint(rawCrc, 16, 16)
	if err != nil {
		return errors.New("cannot convert CRC to integer")
	}

	calculatedCrc := wire.Calculate([]byte(body))
	if calculatedCrc != uint16(receivedCrc) {
		return fmt.Errorf("invalid CRC, received: %04X, calculated: %04X", receivedCrc, calculatedCrc)
	}

	body = strings.TrimSuffix(body, ";")
	parts := strings.Split(body, ";")
	if len(parts) != 2 {
		return errors.New("invalid packet, should have 2 parts")
	}

	rawTimestamp, err := strconv.ParseInt(parts[0], 10, 64)
	if err != nil {
		return errors.New("cannot convert timestamp to integer")
	}
	p.Timestamp = time.Unix(rawTimestamp, 0)
	p.TripId = parts[1]

	return nil
}

// ToPacket converts a TsPacket to its wire representation.
func (p *TsPacket) ToPacket() *string {
	content := fmt.Sprintf("%d;%s;", p.Timestamp.Unix(), p.TripId)
	crc := wire.Calculate([]byte(content))
	result := fmt.Sprintf("<Ts>%s%04X</Ts>", content, crc)
	return &result
}
